File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        #base implementation idea : 
        # store the num in a hashmap as num:freq 
        # check the max freq and return first k num

        res = defaultdict(int)
        lis = []

        for num in nums: #O(n) 
            if num in res:
                res[num] += 1
            else: res[num] = 1
        

        return sorted(res, key=res.get, reverse=True)[:k]
        #this return statement returns k keys with highest value

        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
#include <iostream>
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> count;
        for (int num : nums){
            count[num] += 1;
        }
    // TO PRINT EA VALUE NEED TO ITERATE OVER
    //     for (const auto& entry : count) {
    //     cout << "Key: " << entry.first << ", Value: " << entry.second << endl;
    // }
        // CREATE BUCKETS with freq:[int1, int2, ..]
        vector<vector<int>> buckets(nums.size()+1);
        for(auto& entry: count){
            buckets[entry.second].push_back(entry.first);
        }
        
        vector<int> res;
        for (int i = buckets.size()-1; i >= 0; --i){
            if(!buckets[i].empty()){
                res.insert(res.end(), buckets[i].begin(), buckets[i].end());
                if(res.size() >= k){
                    res.resize(k);
                    break;
                }
            }
        }
        return res;
    }
};